sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -540,22 +546,39 @@ public virtual ServiceBusReceivedMessage AmqpMessageToSBMessage(AmqpMessage amqp
                 }
             }
 
+            return annotatedMessage;
+        }
+
+        public virtual ServiceBusReceivedMessage AmqpMessageToSBReceivedMessage(AmqpMessage amqpMessage, bool isPeeked = false)
+        {
+            AmqpAnnotatedMessage annotatedMessage = AmqpMessageToAnnotatedMessage(amqpMessage, isPeeked);
+
+            ServiceBusReceivedMessage sbMessage = new ServiceBusReceivedMessage(annotatedMessage);
+
             // lock token
 
-            if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
+            sbMessage.LockTokenGuid = ParseGuidBytes(amqpMessage.DeliveryTag);
+
+            amqpMessage.Dispose();
+
+            return sbMessage;
+        }
+
+        public virtual Guid ParseGuidBytes(ReadOnlyMemory<byte> bytes)
+        {
+            if (bytes.Length == GuidSizeInBytes)
             {
                 Span<byte> guidBytes = stackalloc byte[GuidSizeInBytes];
-                amqpMessage.DeliveryTag.AsSpan().CopyTo(guidBytes);
+                bytes.Span.CopyTo(guidBytes);
                 if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
                 {
                     lockTokenGuid = new Guid(guidBytes.ToArray());
```

Comment:
I don't understand this code: TryRead return false if the size of the byte array does not correspond to the size of the Guid struct. But the Guid ctor pretty much does the same. In fact, the Guid ctor on little-endian machines just calls MemoryMarsha.Read. And so, when would the ctor call succeed when the TryRead fails (returns false)?

Reply:
Good question - this code was not actually changed other than doing a slight refactor. I will see if this can be simplified.

Reply:
/cc @jsquire @danielmarbach in case they have additional context.

Reply:
I have no additional insight on this one.   Since we've copied to a temporary buffer already, I'm not sure what the marshal call adds for us here.